teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+
+Today, there is no standardized way to restrict parameter values in Tekton. `Task` authors may have to write an extra input validation `step`, or to embed the input validation logic in the scripts to achieve this validation against the user input. There are two problems with this:
+- **Usability issue**: Task authors are burdened with the task of implementing validation logic, which can lead to errors and inconsistencies.
+- **Error handling issue**: The errors caused by the invalid `param` values can only be caught at `pod` run time.
+
+Therefore, a built-in mechanism is desired to allow `Task`/`Pipeline` authors to declare upfront a list of allowed values for a parameter at authoring time. 
+
+In addition, this feature will enable Tekton to become a step closer to [SLSA hermeticity requirement](https://slsa.dev/spec/v0.1/requirements#hermetic), which requires that all inputs to a build must be fully declared up front. 
+
+
+### Goals
+- Design a Tekton built-in `param` input validation mechanism that constrains the user-provided `string param` value to a set of allowed constants predefined by the author.
+
+### Non-Goals
+- Meeting [SLSA hermeticity](https://slsa.dev/spec/v0.1/requirements#hermetic) requirement.
+
+### Use Cases  
+- As the [buildah](https://github.com/tektoncd/catalog/tree/main/task/buildah/0.5) Task Author, I want to specify the valid values (`oci` or `docker`) of the `param` [`FORMAT`](https://github.com/tektoncd/catalog/blob/4754df9c260e8f102616170175cb132cfc3c5544/task/buildah/0.5/buildah.yaml#L42C1-L44C19). I want to fail the validation if the input is **NOT** valid.
+
+  Without input validation, the buildah task will fail at `pod` runtime due to the invalid `FORMAT` value. The error message is hidden in the container log:
+  
+    ```yaml
+    apiVersion: tekton.dev/v1
+    kind: TaskRun
+    name: buildah-run
+    spec:
+      taskRef:
+        name: buildah
+      params:
+        - name: FORMAT
+          value: "invalid-val"
+      ...
+    ```
+    
+    ```bash
+    # the execution of the above task fails with log:
+    $ kubectl logs buildah-run-pod -c step-build
+    unrecognized image type "invalid-val"
+    ```
+    
+  As a workaround, the `Task` Author must add a validation step before building the image using buildah:
+
+  ```yaml
+  apiVersion: tekton.dev/v1beta1
+  kind: Task
+  metadata:
+    name: buildah
+  spec:
+    params:
+    - name: FORMAT
+      description: The format of the built container, oci or docker
+      default: "oci"
+    ....
+    steps:
+      - name: validate-format-param
+        image: bash
+        script: |
+          # a script to check if $(params.FORMAT) is either `oci` or `docker`.
```

Comment:
alternatively this can also be guarded using whenExpressions when used in a pipeline.

```yaml
when:
  - input: $(params.FORMAT)
     operator: in
     values: ["oci", "docker"]

Reply:
Yes 👍 ! And I think there are some difference between the two
- The Param Enum in Task can be reused in any Pipelines
- Syntax is more concise 
- The behaviour of dependent PipelineTasks are different between [skipping due to when expression](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#guarding-a-task-only) and Param enum

I will add these differences in the TEP

Reply:
Alternative added to TEP